sdk/python/feast/infra/offline_stores/bigquery.py

```diff
@@ -287,23 +291,36 @@ def to_arrow(self) -> pyarrow.Table:
 
 
 def block_until_done(client, bq_job):
-    def _is_done(job_id):
-        return client.get_job(job_id).state in ["PENDING", "RUNNING"]
+    """
+    Waits a maximum of 30 minutes for bq_job to finish running.
+    Args:
+        client: A bigquery.client.Client to monitor the bq_job.
+        bq_job: The bigquery.job.QueryJob that blocks until done runnning.
+    Raises:
+        BigQueryJobStillRunning exception if the function has blocked longer than 30 minutes.
+        BigQueryJobCancelled exception on a KeyboardInterrupt.
+    """
 
-    @retry(wait=wait_fixed(10), stop=stop_after_delay(1800), reraise=True)
+    @retry(
+        wait=wait_fixed(10),
+        stop=stop_after_delay(1800),
+        retry=retry_if_exception_type(BigQueryJobStillRunning),
+        reraise=True,
+    )
     def _wait_until_done(job_id):
-        return _is_done(job_id)
+        if client.get_job(job_id).state in ["PENDING", "RUNNING"]:
+            raise BigQueryJobStillRunning(job_id=job_id)
 
     job_id = bq_job.job_id
-    _wait_until_done(job_id=job_id)
+    try:
+        _wait_until_done(job_id=job_id)
+    except KeyboardInterrupt:
```

Comment:
Just wondering, would it make more sense to have this in a `finally` block instead of a `except KeyboardInterrupt`? that may also handle sigterm/sigkill.

Reply:
Good call. We could have a finally statement that send an asynchronous call to cancel the BQ job in case something goes wrong 👍

Reply:
Added the `finally` clause to catch other needs for cancellation (including timeout and `KeyboardInterrupt`).